### DISCUSSION TITLE
Best convolutional layers when working with 3D meshes

### DISCUSSION THREAD

**Discussion by JJab0n** [pyg-team/pytorch_geometric#4813] [Q&A] [answered]
Hello world,

I am currently using pytorch geometric to develop generative models that work with 3D objects. Each 3D object (**stl** file) is represented by a triangular mesh that is converted to a graph, including node features (x,y,z coordinates) and edge index (connection between nodes). 

I have been able to code a Graph Variational Autoencoder following some recent articles [1,2] that can parameterizes and reconstructs each mesh of the dataset. In this GVAE I have used the Chebyshev spectral graph convolutional operator (**ChebConv**) from [3] currently implemented in Pytorch Geometric. 

Although the results obtained are not bad at all, I consider the reconstruction error is still high. For this reason, I would like to ask if there are other convolutional layers that are more suitable when working with meshes of 3D objects. Maybe @rusty1s knows projects which use other options.

Thanks in advance.

References:
[1] [Mesh Variational Autoencoders with Edge Contraction Pooling](
https://doi.org/10.48550/arXiv.1908.02507)
[2] [Generating 3D Faces using Convolutional Mesh Autoencoders](https://doi.org/10.48550/arXiv.1807.10267)
[3] [Convolutional Neural Networks on Graphs with Fast Localized Spectral Filtering](https://doi.org/10.48550/arXiv.1606.09375)

**Comment by rusty1s** [comment, marked as answer]
Yes, there are many options which should work way better than `ChebConv` on this kind of data, e.g.:
* `PointConv`
* `SplineConv`
* `GMMConv`
* `EdgeConv`
The key difference in these approaches is that spatial coordinates of points/nodes are moved to spatial relations as edge features. This way, the model is translation-invariant and generalized better to unseen positions.

**Reply by JJab0n** [reply]
Thank you for your guidelines. I really appreciate it. 

Regarding **PointConv** and **EdgeConv**, everything is clear. I have implemented it, but I still need to find the best internal configuration of neural networks for each convolutional layer. 

However, I have a question with respect to **SplineConv** and **GMMConv**. They both need edge features, which I have calculated using Cartesian transform from Pytorch Geometric due to the fact that my original graph is defined using only node features and edge indices. Is this the best way to create a complete graph (_x_, _edge_index_ and _edge_attr_) from a .stl object? Should I follow another method (Euclidean distance, Polar, etc) or even another alternative to represent the source mesh?

**Reply by rusty1s** [reply]
What we found out is that `Cartesian` usually works best across a variety of datasets. If your model desperately needs to be rotation invariant, you can also look at the `PointPairFeatures` transform, which is a rotation-invariant method to relate two points to each other.
